astropy/io/fits/header.py

```diff
@@ -599,7 +605,7 @@ def _from_blocks(cls, block_iter, is_binary, sep, endcard, padding):
             if not is_binary:
                 block = encode_ascii(block)
 
-        if not end_found and is_eof and endcard:
+        if not end_found and is_eof and endcard and is_header:
```

Comment:
This is a perfect interpretation of what I wrote in #4351, though I think what I wrote there is not quite correct.  Normally `SIMPLE =` is only in the first header of a FITS file.  So if we were dealing with something like an accidentally truncated FITS file, the more likely thing you would see at the beginning of another header would be `XTENSION =`.

But I think I found a simpler solution to this problem.  Instead of what you did, try moving the lines a bit below this:

```python
        header_str = ''.join(read_blocks)
        _check_padding(header_str, actual_block_size, is_eof,
                       check_block_size=padding)
```

to just before this line.  As far as I can tell that also solves the problem, and doesn't break any existing tests.  Please try out that solution and double-check me.

Reply:
> Normally `SIMPLE =` is only in the first header of a FITS file. So if we were dealing with something like an accidentally truncated FITS file, the more likely thing you would see at the beginning of another header would be `XTENSION =`.

You are right. I should have thought about that! Anyway, moving the check before the if statement works like a charm.